### PR TITLE
feature: resource generation with fields identification and model

### PR DESCRIPTION
**Pull request by mihaimdm22**
User model and Admin::User model contain following columns: 
```
"id"=>:integer, 
"email"=>:string, 
"first_name"=>:string, 
"last_name"=>:string, 
"roles"=>:json, 
"birthday"=>:date, 
"custom_css"=>:text, 
"team_id"=>:integer, 
"encrypted_password"=>:string, 
"reset_password_token"=>:string, 
"reset_password_sent_at"=>:datetime, 
"remember_created_at"=>:datetime, 
"created_at"=>:datetime, 
"updated_at"=>:datetime, 
"active"=>:boolean
```

1. Running `rails g avo:resource User` or `rails g avo:resource User --generate-fields false` will generate you:
```
module Avo
  module Resources
    class User < Resource
      def initialize
        @title = :id
        @search = :id   
      end

      fields do
        id   
      end
    end
  end
end
```

2. Running `rails g avo:resource User --generate-fields` or `rails g avo:resource User --generate-fields true` or `rails g avo:resource User --generate-fields anythingelsestring` will generate you:
```
module Avo
  module Resources
    class User < Resource
      def initialize
        @title = :id
        @search = :id
      end

      fields do
        id
        text :email
        text :first_name
        text :last_name
        code :roles
        date :birthday
        textarea :custom_css
        number :team_id
        password :encrypted_password
        text :reset_password_token
        datetime :reset_password_sent_at
        datetime :remember_created_at
        datetime :created_at
        datetime :updated_at
        boolean :active
      end
    end
  end
end
```

3. Running `rails g avo:resource User --model_class Admin::User` or `rails g avo:resource User --generate-fields false --model_class Admin::User` will generate you:
```
module Avo
  module Resources
    class User < Resource
      def initialize
        @title = :id
        @search = :id
        @model = Admin::Use
      end

      fields do
        id
      end
    end
  end
end
```

4. Running `rails g avo:resource User --generate-fields --model_class Admin::User` or `rails g avo:resource User --generate-fields true --model_class Admin::User` will generate you:
```
module Avo
  module Resources
    class User < Resource
      def initialize
        @title = :id
        @search = :id
        @model = Admin::Use
      end

      fields do
        id
        text :email
        text :first_name
        text :last_name
        code :roles
        date :birthday
        textarea :custom_css
        number :team_id
        password :encrypted_password
        text :reset_password_token
        datetime :reset_password_sent_at
        datetime :remember_created_at
        datetime :created_at
        datetime :updated_at
        boolean :active
      end
    end
  end
end
```

5. Running `rails g avo:resource User description:trix members:integer birthday:number --generate_fields` will generate you:
```
module Avo
  module Resources
    class User < Resource
      def initialize
        @title = :id
        @search = :id      
      end

      fields do
        id
        text :email
        text :first_name
        text :last_name
        code :roles
        number :birthday
        textarea :custom_css
        number :team_id
        password :encrypted_password
        text :reset_password_token
        datetime :reset_password_sent_at
        datetime :remember_created_at
        datetime :created_at
        datetime :updated_at
        boolean :active
        trix :description
        number :members
      end
    end
  end
end
```

!!!
Running `rails generate avo:resource --help` will output in console:
```
Usage:
  rails generate avo:resource NAME [key:value] [options]

Options:
  [--skip-namespace], [--no-skip-namespace]    # Skip namespace (affects only isolated applications)
  [--generate-fields], [--no-generate-fields]  # Looks for fields in the model and generates them.
  [--model-class=MODEL-CLASS]                  # The name of the model.

Runtime options:
  -f, [--force]                    # Overwrite files that already exist
  -p, [--pretend], [--no-pretend]  # Run but do not make any changes
  -q, [--quiet], [--no-quiet]      # Suppress status output
  -s, [--skip], [--no-skip]        # Skip files that already exist

Creates an Avo resource
```

### CODE REVIEW

File: lib/generators/avo/resource_generator.rb [hunk 1]
```diff
@@ -1,9 +1,166 @@
...
+        email: {
+          field: 'text',
+        },
+        password: {
+          field: 'password',
+        },
+        password_confirmation: {
+          field: 'password',
+        },
+        stage: {
```
Comment (adrianthedev) [inline lib/generators/avo/resource_generator.rb:150]: IDK if we should generate this as a select. no options given.

We can use `models.defined_enums` to get all the enums ans set them as a select with `enum` option.
ex: `select :stage, enum: ::Project.stages`

File: lib/generators/avo/resource_generator.rb [hunk 2]
```diff
@@ -1,9 +1,166 @@
...
+        references: {
+          field: 'belongs_to',
+        },
+        json: {
+          field: 'code',
+        },
+      }
+
+      name_mappings = {
+        id: {
```
Comment (adrianthedev) [inline lib/generators/avo/resource_generator.rb:132]: We can have a `country` field.

```suggestion
		country: {
			field: 'country',
		},
        id: {
```

File: lib/generators/avo/resource_generator.rb [hunk 3]
```diff
@@ -1,9 +1,166 @@
...
+      fields = fields.except('id') if fields.key?('id')
+
+      fields
+    end
+
+    def generated_params
+      columns_with_type = {}
+      begin
+        model = @model_class.classify.constantize
+        model.columns_hash.each { |k, v| columns_with_type[k] = v.type }
```
Comment (adrianthedev) [inline lib/generators/avo/resource_generator.rb:35]:
```suggestion
        columns_with_type = model.columns_hash.map { |k, v| [columns_with_type[k], v.type] }.to_h
```

File: lib/generators/avo/resource_generator.rb [hunk 4]
```diff
@@ -1,9 +1,166 @@
...
+      rescue NameError => e
+        puts 'Name error occurs. There is no ' + @model_class.classify + ' model.'
+      rescue => e
+        puts 'Other error occured.'
+      end
+
+      result = {}
+      if !columns_with_type.empty?
```
Comment (adrianthedev) [inline lib/generators/avo/resource_generator.rb:43]:
```suggestion
      if columns_with_type.present?
```

File: lib/generators/avo/resource_generator.rb [hunk 5]
```diff
@@ -1,9 +1,166 @@
...
+        columns_with_type.each do |fieldname, fieldtype|
```
Comment (adrianthedev) [inline lib/generators/avo/resource_generator.rb:44]: They are two words.
```suggestion
        columns_with_type.each do |field_name, field_type|
```

**Review by adrianthedev — CHANGES_REQUESTED** [review]
Some notes:

- we can get the attachment/attachments through `model.reflections`
- we can get the relationships trhough `model.reflections`
- ignore some fields:
  - default timestamps `created_at`, `updated_at`
  - from devise `reset_password_sent_at`, `remember_created_at`, `reset_password_token`


On `rails g avo:resource User --generate-fields --model_class Admin::User` I get an extra space.

image

Same on `rails g avo:resource User --generate-fields`

image

File: lib/generators/avo/resource_generator.rb (hunk 3 above)
Comment (mihaimdm22) [reply lib/generators/avo/resource_generator.rb:35]: doesn't work this way

**Comment by IrvanFza** [comment]
Hi @mihaimdm22 @adrianthedev, just wondering if this feature will be implemented on Avo any time soon. Really love your work on Avo, and probably will become the game-changer on Rails admin dashboard-related gem. This feature will make Avo more awesome!

**Comment by adrianthedev** [comment]
Hey @IrvanFza 
We don't have the capacity to tackle this feature now, but we're more than open if someone would like to take this PR over. We're pretty supportive of our contributors and help them navigate the codebase.

**Pull request closed without merging**